### PR TITLE
#1946 Convert Document, Content, DocumentLoader to interfaces

### PR DESCRIPTION
**Pull request by kpavlov**
## Issue
Explains #1946

## Change

Refactor following classes to interfaces for better modularity and readability:
- `Document` to interface and introduce `SimpleDocument` as a record implementation
- `Content` to interface and introduce `TextContent` as a record implementation
- `DocumentLoader` to interface
- `DocumentParser` to use generics

**This is a potentially breaking change,** because public constructors of converted classes were removed. Luckily, static factory methods had been used already in the code.

## General checklist

- [ ] There are no breaking changes
- [x] I have added unit and integration tests for my change
- [ ] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [x] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for adding new maven module

- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`

## Checklist for adding new embedding store integration

- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration

- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

File: langchain4j-core/src/main/java/dev/langchain4j/data/document/DocumentLoader.java [hunk 1]
```diff
@@ -5,10 +5,7 @@
 /**
  * Utility class for loading documents.
  */
-public class DocumentLoader {
-
-    private DocumentLoader() {
-    }
+public interface DocumentLoader {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/data/document/DocumentLoader.java:8]: Why is this change required?

File: langchain4j-core/src/main/java/dev/langchain4j/data/document/DocumentParser.java [hunk 2]
```diff
@@ -6,7 +6,7 @@
  * Defines the interface for parsing an {@link InputStream} into a {@link Document}.
  * Different document types require specialized parsing logic.
  */
-public interface DocumentParser {
+public interface DocumentParser<T extends Document> {
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/data/document/DocumentParser.java:9]: Why is this required?

File: langchain4j-core/src/main/java/dev/langchain4j/data/document/DocumentLoader.java [hunk 3]
```diff
@@ -20,10 +17,10 @@ private DocumentLoader() {
      * @return The loaded document.
      * @throws BlankDocumentException when the parsed {@link Document} is blank/empty.
      */
-    public static Document load(DocumentSource source, DocumentParser parser) {
+    static Document load(DocumentSource source, DocumentParser parser) {
         try (InputStream inputStream = source.inputStream()) {
             Document document = parser.parse(inputStream);
-            source.metadata().asMap().forEach((key, value) -> document.metadata().add(key, value));
+            source.metadata().toMap().forEach((key, value) -> document.metadata().add(key, value));
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/data/document/DocumentLoader.java:23]: We need to be careful with this change, we can break something

**Review by ghost — COMMENTED** [review]
@kpavlov thank you!

File: langchain4j-core/src/main/java/dev/langchain4j/data/document/DocumentLoader.java (hunk 1 above)
Comment (kpavlov) [reply langchain4j-core/src/main/java/dev/langchain4j/data/document/DocumentLoader.java:8]: Because there is no interface for DocumentLoaders and now e.g. `FileSystemDocumentLoader` or`AmazonS3DocumentLoader` can implement the at least marker interface.
It would be even better to implement :

```java
interface DocumentLoader<T extends Document> {
   T loadDocument(DocumentSource<T> source);
}
```

Spring AI has a similar concept of [DocumentReader](https://github.com/spring-projects/spring-ai/blob/main/spring-ai-core/src/main/java/org/springframework/ai/document/DocumentReader.java)

File: langchain4j-core/src/main/java/dev/langchain4j/data/document/DocumentLoader.java (hunk 3 above)
Comment (kpavlov) [reply langchain4j-core/src/main/java/dev/langchain4j/data/document/DocumentLoader.java:23]: I read the deprecation comment, and go ahead with this change, hoping that tests can verify that. I hope this should be fine as long as unit and integration tests are green.
But I agree, this needs care, because Object to String conversion may need a special care

File: langchain4j-core/src/main/java/dev/langchain4j/data/document/DocumentParser.java (hunk 2 above)
Comment (kpavlov) [reply langchain4j-core/src/main/java/dev/langchain4j/data/document/DocumentParser.java:9]: To introduce different types of document in the future.

**Comment by kpavlov** [comment]
@langchain4j , would it be fine to start with just the first [commit](https://github.com/langchain4j/langchain4j/pull/1959/commits/b7ce3543a9bcabbf4aafc4dfd4a57fc4d5df77db)? I can revert other commits or create a new PR
It make sense to refactor DocumentLoaders separately

File: langchain4j-core/src/main/java/dev/langchain4j/data/document/DocumentLoader.java (hunk 1 above)
Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/data/document/DocumentLoader.java:8]: @kpavlov loaders were left out without a common interface on purpose, as they all have different APIs and it is hard to find a common one without forcing a lot of limitations. Also loaders usually are used in the beginning of the pipeline, so it is not really required for them to be polymorphic.

I am OK with a marker interface solely for discoverability, but I would not force a common API.
BTW, users are not supposed to use `DocumentSource` directly, they are supposed to use `XyzDocumentLoader` which will then deal with `DocumentSource`s internally.

File: langchain4j-core/src/main/java/dev/langchain4j/data/document/DocumentParser.java (hunk 2 above)
Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/data/document/DocumentParser.java:9]: Could you please elaborate? What other types do you foresee?

File: langchain4j-core/src/main/java/dev/langchain4j/data/document/DocumentLoader.java (hunk 3 above)
Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/data/document/DocumentLoader.java:23]: Comment for myself: check https://github.com/langchain4j/langchain4j/issues/1496

**Comment by kpavlov** [comment]
@langchain4j , let's take  #1988
